rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/Negotiation.cpp

```diff
@@ -858,8 +888,14 @@ class Negotiation::Implementation
         approvals.erase(approval_callback_it);
     }
 
+    if (conclusion_callback)
+      conclusion_callback(msg.conflict_version, msg.resolved);
+    
     // Erase these entries because the negotiation has concluded
-    negotiations.erase(negotiate_it);
+    // NOTE(ddengster): retained_history_count is used for testing 
+    // and doesnt have full functionality
+    if (negotiations.size() > retained_history_count)
```

Comment:
This implementation won't really do what the API suggests. It will save the first `N` negotiations that ever occur, and then discard all the rest as they finish.

Instead of this, I would recommend adding a field to the `Implementation`:

```
std::map<Version, rmf_traffic::schedule::Negotiation> history;
```

to store the finished negotiations. Then here you can do:

```
if (retained_history_count > 0)
{
  while (history.size() >= retained_history_count)
    history.erase(history.begin());

  history.insert({msg.conflict_version, std::move(negotiate_it->room.negotiation)});
}
```

The `std::move` semantics will be important here because the `rmf_traffic::schedule::Negotiation` class cannot be copied.

After this, you can tweak the implementation of `table_view(~)` so that it checks in `history` if the negotiation can't be found in `negotiations`.

Reply:
yea I put it there as an aid for testing, didn't fully implement it, hence the NOTE (it should really be a TODO)

also, i believe std::map does not preserve the order of insertion? std::deque would be a better alternative

Reply:
> std::deque would be a better alternative.

I considered `std::deque`, but then looking up a specific negotiation has O(N) complexity.

> i believe std::map does not preserve the order of insertion? 

`std::map` does preserve the order of the `Version` value, which will reflect the order that the negotiations were created in. So the history would contain the last `N` completed negotiations based on their time of creation rather than their time of completion. I think that's a reasonable compromise between complexity and intent.

> Mind if I put it these changes into a different PR?

I think the changes should be straightforward enough to get them into this PR. It would just be the code I put in my last comment plus maybe two more lines in `table_view(~)`. We're expected to maintain a high level of quality control for `rmf_core`, so I'd be reluctant to merge a feature into `master` that doesn't do what it describes.

Reply:
> I considered std::deque, but then looking up a specific negotiation would have O(N) complexity.

This is a feature to aid testing, and I always intended N to be small(<10). But it's fine, we can switch to map.

Reply:
reworked e3f0902